src/python/pants/core/target_types.py

```diff
@@ -87,14 +87,39 @@ def __init__(self, url: str, *, len: int, sha256: str, filename: str = ""):
 
         if not self.filename:
             raise ValueError(
-                "Couldn't deduce filename from `url`. Please specify the `filename` argument."
+                softwrap(
+                    f"""
+                    Couldn't deduce filename from `url`: '{self.url}'.
+
+                    Please specify the `filename` argument.
+                    """
+                )
             )
         if "\\" in self.filename or "/" in self.filename:
-            raise ValueError("`filename` cannot contain a path separator.")
+            raise ValueError(
+                f"`filename` cannot contain a path separator, but was set to '{self.filename}'"
+            )
 
 
 class AssetSourceField(SingleSourceField):
     value: str | HTTPSource  # type: ignore[assignment]
+    # @TODO: Don't document http_source, link to it once https://github.com/pantsbuild/pants/issues/14832
+    # is implemented.
+    help = softwrap(
+        """
+        The source of this target.
+
+        If a string is provided, represents a path is relative to the BUILD file's directory,
+        e.g. `source='example.ext'`.
+
+        If an http_source is provided, represents the network location to download the source from.
+        `http_source` has the following signature:
+            http_source(url: str, *, len: int, sha256: str, filename: Optional[str])
```

Comment:
FYI our `help` system converts `Optional[str]` to `str | None`, so better to use that here too.
```suggestion
            http_source(url: str, *, len: int, sha256: str, filename: str | None)
```

Reply:
But also maybe it's more accurate to use `:str = ""` - it's an error if someone puts `None`

Reply:
Ahh fair